README.md

```diff
@@ -781,6 +781,48 @@ nestingLimit: 6
 
 This will prevent the user from nesting tags more than 6 levels deep. Tags deeper than that are stripped out exactly as if they were disallowed. Note that this means text is preserved in the usual ways where appropriate.
 
+### Advanced filtering
+
+For more advanced filtering you can hook directly into the parsing process using tag open and tag close events.
+
+The `onOpenTag` event is triggered when an opening tag is encountered. It has two arguments:
+- `tagName`: The name of the tag.
+- `attribs`: An object containing the tag's attributes, e.g. `{ src: "/path/to/tux.png" }`.
+
+The `onCloseTag` event is triggered when a closing tag is encountered. It has the following arguments:
+- `tagName`: The name of the tag.
+- `isImplied`: A boolean indicating whether the closing tag is implied (e.g. `<p>foo<p>bar`) or explicit (e.g. `<p>foo</p><p>bar</p>`).
+
+For example, you may want to add spaces around a removed tag, like this:
+```js
+const allowedTags = [ 'b' ];
+let addSpace = false;
```

Comment:
This example might struggle with nested tags in some situations, because `addSpace` will get reset and then it doesn't get "popped" back to its previous value when we return to the enclosing tag's context. I think that could be overcome in the example without changing your API, but that depends on how the API will handle tags that do not need or have a closing tag like `link`, do they fire only `onOpenTag`? If so it would be very difficult to get this 100% right. What are your thoughts on this?

Reply:
Could you give an example HTML snippet you have in mind? Would be easier for me to reason about it.

Regarding the example, this is not necessarily meant to be 100% production-ready and cover all possible edge cases, but rather to illustrate new possibilities that open up with these event handlers. In a real-life scenario people may want to have a special list of tags or custom rules where this applies (e.g. only block-level tags, only visible tags, a subset of these etc.)